### ISSUE TITLE
CD-ROM Support

### ISSUE THREAD

**Issue by SolraBizna** [ayrtonm/psx-sdk-rs#3]
Okay, so. I got everything else squared away in my dev environment, and `println!` and friends were recently changed in such a way that (coincidentally?) caused them to start working with Mednafen's BIOS output hack... so the next obstacle in my project is CD-ROM IO support.

The current wrapping of the BIOS file APIs only works for memory cards. In particular, since memory cards can only be read and written in whole blocks of 128 bytes, this is baked into the API in the form of taking `[[u8; 128]]` as parameters (slices of 128-byte arrays). What's the best way forward on this?

1. Change the API to take `[u8]` and give up on static length checking.
    i. Add a runtime length check, so that an IO on a memory card file that isn't a multiple of 128 bytes, or an IO on a CD-ROM file that isn't a multiple of 2048 bytes, still gets a legible error.
    ii. Leave out the length check, and let the BIOS signal an error if the user gets the length wrong. (At least, I assume it signals an error. I haven't gotten that far in my testing. See below...)
2. Make separate wrappers, one for `bu*:` files, one for `cdrom:` files. Make the CD wrapper take `[[u8; 2048]]`. [ETA: since limited `const_generics` has landed, the block size could be a template parameter instead of needing a separate struct.]

I'd personally be happier if it were simple byte slices, with or without a length check. If I know, for example, that my game's save data is exactly 384 bytes long, it seems inconvenient to have to split that into three 128 byte arrays. A runtime length check does potentially incur a small runtime cost, but that should be insignificant in the face of the overhead of the actual IO. (Is there a crate that I'm not aware of that sugar-coats the slice-of-arrays approach?)

Related to CD-ROM... while poking around, trying to learn more about how the BIOS handles `cdrom:` IO, I ran into a strange issue. Any call I make that involves CD IO hangs the system. In particular, `open_file("cdrom:\\DIR\\FILE.EXT;1", 1)` with a CD in the drive quickly goes into a busy loop on `TestEvent(0xF1000001)`. Tho no$psx spec says that this event is an "Event Event", and also says "not used by BIOS; maybe used by Sony's devkit?" I have to assume that there's some sort of event/IRQ state that needs to be set up and isn't, or that is being corrupted by the boot process. I tried adding a call to `kernel::CdInit()` with no effect (and no$ says it's not needed anyway). I tried doing `kernel::init_pad` and `kernel::start_pad`, the latter of which hurls a `SystemErrorUnresolvedException`—but only if there's a CD in the drive?

I've confirmed that the BIOS call is going through as intended by single-stepping with Mednafen's debugger, including the zero-termination and pointer cast that I glossed away above. And I'm fairly sure my CD is readable, since the BIOS is able to successfully boot my EXE in the first place.

I'm stumped...

**Comment by ayrtonm** [comment]
> The current wrapping of the BIOS file APIs only works for memory cards.
> What's the best way forward on this?

I think it's best to make separate wrappers since `OpenOptions` and a lot of memcard methods don't apply to CD files (but `&[[u8; 128]]` could definitely be improved too). It's mostly obvious which methods don't apply to CDs, but I'd rather get a compile-time error for forgetting one of those small details. To reuse code for shared behavior I'd probably do `File<T: FileType>` where `FileType` is just a marker trait. We could also consider using macros or making `File` a trait if that makes more sense.


> I'd personally be happier if it were simple byte slices, with or without a length check

Yeah, I'm also not very content with the current solution. Even if `File` were only for memcards, it's not very ergonomic like you point out.

> let the BIOS signal an error if the user gets the length wrong

Passing an incorrectly sized `&[u8]` returns `Ok(0)` instead of an error which tripped me up since I initially missed the note about writing multiples of 128 bytes. I considered changing this case to `Err`, but I wasn't a fan since it sort of changes the return code's meaning. That's why I used `&[[u8; 128]]` since its memory layout is the same as any valid input to those methods so it's basically a convenient "generic" newtype. If we were to keep using a newtype here, a better option might be to wrap `&[[u8; 128]]` in a struct and implement the `Index` trait to allow using it like a normal array.

A simpler solution might be the (incomplete) feature [const_evaluatable_checked](https://github.com/rust-lang/rust/issues/76560) to allow `fn write<const N: usize>(src: &[u8; 128 * N])`. On one hand it provides static length checks. The downsides is that each call site becomes length-specific. Also the const parameter can't be inferred (yet). Omitting the parameter doesn't give a very helpful message, but specifying an `N` that doesn't match the argument gives a clear compile-time error pointing out the argument's length so only the first problem is a real issue. What do you think about this idea?

> Related to CD-ROM...

I haven't looked at the CD functions in months, but I'll get around to helping you debug this at some point.

**Comment by SolraBizna** [comment]
With types named like `CdromFile` and `MemCardFile`, it's more obvious at a glance that these are low-level, domain-specific primitives. The standard library's `File` struct implements `Read`, `Write`, and `Seek` traits; probably best to take an approach like that here as well. As for statically-checked IO, you could have `BlockRead<N>` for example to allow opting into that, while having more generic `Read`/`Write` implementations that take byte slices and involve more runtime checking.

Returning `Ok(0)` because the provided length was shorter than one block would definitely be surprising. Does it also return `Ok(0)` for non-block-aligned sizes larger than one block? If so... wow. (It would be nowhere near the most alarming thing about this BIOS, though...) Either way, reading into a buffer that's too small to *ever possibly succeed* is a case we can detect, and probably should have a special error for. If the BIOS rounds down to the nearest block, instead of throwing up its hands and giving up if the read is not *exactly* block-aligned (which the PlayStation's hastily-written BIOS very well might do, for all I know), then a `< BLOCK_SIZE` check is the only one we would need to add. No matter what, we should ensure that `Ok(0)` always mean EOF, and never that the read failed for some other reason. That's my opinion, at least.

`const_evaluatable_checked` gets us most of the way to an ergonomic, statically-checked solution. Lack of inference of `N` is a pain, but a little bit of turbofish never hurt anybody. Multiple call sites based on length is also probably not too serious a problem, since the function will probably usually be inlined. Not being able to have a dynamic length for such a call sucks, but if byte slices are *also* an option, then that's not such a big deal.

Another approach to consider would be having low-level, *single-block oriented* IO types `CdromFile` and `MemCardFile`, implementing block-oriented IO traits (`BlockRead<N>`/etc.?); and a higher-level `File<N>` type that implements byte-slice-oriented IO traits (`Read`/etc.) *and has a built-in one-block buffer* (of `N` bytes), doing partial reads and writes and such as needed. This way, optimization-crazies (like me) could use the low level types and never copy data twice, and lazies (also like me) could just read/write arbitrary sizes and let the SDK sort it out. And since Rust has dead code elimination as a core language feature instead of an optional optimization, everybody wins!

**Comment by ayrtonm** [comment]
> File struct implements Read, Write, and Seek traits; probably best to take an approach like that

I intentionally avoided that approach since it doesn't make much sense for the psx. In `std` separate traits for each method makes sense since they have many different implementors, but in this case they'd each have a max of 4 implementors (combinations of memcard/CD-R and BIOS/IO registers). Also only `Seek` could really be implemented by all four since `Read` would be block-size specific. I think importing traits to use their methods is a bit annoying and even more so if they're not providing much benefit.

I do see a benefit in making traits for each file type though. Including all methods in these traits would allow generic functions that work with files of a given type managed by either the BIOS or directly with memory-mapped I/O. I can't think of a good use-case for a trait covering both memcards and CD-Rs, but let me know if you have something in mind.

For code reuse in the BIOS fs implementation we could parameterize the `File` type or make an internal `BIOSFile` trait (specifically for code reuse) with defaults. In this case I think parametrized types makes more sense since
```
impl<FileTy> File<FileTy> {
    fn read() { }
    fn seek() { }
}

impl File<MemCard> {
    fn write () { }
}
```
seems clearer (and produces more readable docs) than including the common functions (`read`, `seek`) in the `BIOSFile` trait then making `write` an inherent method for the memcard file type. Also in this example they're inherent impls, but it could work with the trait per filetype idea by renaming the functions above to`internal_*` and having the filetype trait functions call them.

> Does it also return Ok(0) for non-block-aligned sizes larger than one block?

yes

I wanted to keep these wrappers simple and avoid runtime checks, but I care more about minimizing surprising behavior here. Right now the two things that may be surprising to someone unfamiliar with this code are

- `read(&mut [u8; 128 * N])` will not compile with a correctly sized slice
- calling `read(&mut [u8])` with an incorrectly sized slice returns `Ok(0)`

The first point rejects valid programs so there definitely has to be a version for slices.

For the second point I'd argue that rounding down to the nearest block would also be surprising behavior, even if it is more like `std`. With rounding I could see a case where a slice's length is miscalculated leading to reading less than expected and the result is only checked for `Ok` and not a specific length (since it'll always return the same size for a given input slice). I also think restricting slice sizes makes the intention more clear to someone unfamiliar with the code. But returning `Ok(0)` is not ok so we could check the length before the BIOS call and return `ErrorKind::SectorSize` in that case.

**Comment by ayrtonm** [comment]
Actually after thinking about how to implement `read(&mut [u8])` again, I think it'll be best to just relabel `Ok(0)` in the return statement to `ErrorKind::SectorSize`. That would avoid the additional check in all cases at the expense of a BIOS call that's bound to fail in the invalid-size slice case (which shouldn't exist in correct programs anyway) and an extra branch in the existing `match`.

Also since the `Ok` result is just the input slice length, the return type could change to `Result<(), E>` (like [`read_exact`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact)) since it's not giving any new info and just complicates the `match` comparison.

**Comment by SolraBizna** [comment]
Let's not forget about short reads due to EOF. If I try to read four blocks and there are only three blocks left, I— ...well, I guess I'm just *assuming* the BIOS will return three blocks read because otherwise its return value is almost totally informationless...

I vaguely remember reading somewhere in the no$psx specs that the BIOS will return a whole block for the end of a file even if that file isn't block-length? I could be remembering that wrong.

An advantage of implementing `Read`/`Write` traits is that they allow adaptation. For example, if our `File` implements `Read` then our caller could use a general-purpose `InflatingReader` adapter around it to decompress what's being read. But that would only be possible for a buffered adapter *around* the BIOS file IO routines, which isn't the current area of implementation. Trying to use a generic adapter around a `Read`/`Write` implementation that chokes if the slice isn't exactly the right size would be a recipe for disaster. So, it's just as you said, in that there's no benefit for doing it for BIOS files. I'd still recommend doing so for a friendly higher-level buffering adapter, though.

As far as naming things, I do think naming the user-visible type something other than just `File` is valuable. `File` might create an expectation that it will act something like the standard `File` type, in particular allowing arbitrary byte-granular IOs; whereas if one sees types `CdromFile` and `MemCardFile` that both implement `BIOSFile`, one will be more likely to assume that pitfalls lie ahead and look more closely.

Oh, and, an important optimization detail. If we have:

```rust
fn some_func(slice: &mut [u8]) {
  if slice.len() % 128 != 0 { panic!("Must be a multiple of 128 bytes!"); }
  // do something with slice
}
```

If a call to `some_func` ends up getting inlined (which in our case it likely will), the compiler will almost always be smart enough to omit the length check if it can prove it's unnecessary. So:

```rust
fn read_and_process_some_data() {
  let mut buf: [0u8; 384];
  some_func(&mut buf[..]);
  // ...
}
```

has no length check in the runtime code at all. It's actually quite smart about this, capable of doing surprisingly complicated reasoning to parse through more complex cases. Therefore, if we do have a slice version of the IO function, and that has a length check written in the code, a lot of the time we won't be incurring any runtime overhead for it.

Oh, but this will be a cross-crate call, so I think that means it's less likely to be inlined without explicitly enabling lto?

Sorry for rambling, it's 5AM here and I'm sick...

**Comment by ayrtonm** [comment]
> Let's not forget about short reads due to EOF. If I try to read four blocks and there are only three blocks left...

This case reads the expected 3 blocks but returns `Ok(4)` and seeking to the very end of the file and reading returns `Err(SectorAlignment)`. "Quirks" like these are why I intended for this to be a bare-bones wrapper instead of supporting every feature the BIOS attempts to implement. I'll do some more tests and take a look at the BIOS function's disassembly, but it'll likely return `Result<()>` and `Ok(0)` will probably be relabeled as `Err`.

> An advantage of implementing Read/Write traits is that they allow adaptation

I agree, but keep in mind that with traits per filetype `MemCard` would effectively be `Write`. I can't imagine a situation here where `T: Write + !Read` or `T: Write + !Seek` is needed. I think `Read` might have some value, but the methods' behavior would depend on the filetype's block-size so it might not be so straightforward to implement `InflatingReader`. Since this module is intended to just wrap the BIOS calls in idiomatic rust instead of recreating every aspect of `std::fs`, I'd say let's avoid traits for now. Nothing really prevents external crates from defining and implementing these traits though.

> Trying to use a generic adapter around a Read/Write implementation that chokes if the slice isn't exactly the right size would be a recipe for disaster

I think my explanation could've been better. This really isn't a problem since `read` with incorrectly-sized slices should _return_ `Err` instead of panicking. It's just a BIOS function call with a relabeled return code wrapped in `Result`. There shouldn't be wildly different behavior from what calling the BIOS from C would do (besides the call to `as_cstr` to allow input `&str` w/o null-termination).

> types CdromFile and MemCardFile that both implement BIOSFile

Like I mentioned earlier, since that trait would cover a subset of one type's methods this produces docs that are hard to follow. To be more concrete, the docs would show `write` under methods for `File<MemCard>` but `read` would be under `File<MemCard>`'s trait implementations and those methods aren't visible by default. Aside from the docs, conceptually it seems a bit strange to have some of the functionality (`write`) for one type of BIOS file outside the `BIOSFile` trait.

> a lot of the time we won't be incurring any runtime overhead for it.

I understand the argument you're making, but I'd rather not add runtime checks based on the hope that the compiler will inline certain things.

> Oh, but this will be a cross-crate call, so I think that means it's less likely to be inlined without explicitly enabling lto?

To answer your question, cross-crate calls are not inlined without LTO. When marked with `#[inline]` (or even `#[inline(always)]`) they [_may_ be inlined](https://doc.rust-lang.org/reference/attributes/codegen.html#the-inline-attribute) without LTO, but are not guaranteed to be.

**Comment by ayrtonm** [comment]
I started a new branch to actually start implementing this. After playing around with `const_evaluatable` I remembered how unstable it is and why I previously removed it after considering it for a different feature. Maybe in the future it'll be usable for static length-checks but for now it's best to just have a single `read(&mut [u8])` that doesn't do explicit length-checking but relabels `Ok(0)` in cases that may be confusing.

**Comment by SolraBizna** [comment]
> > Let's not forget about short reads due to EOF. If I try to read four blocks and there are only three blocks left...
> 
> This case reads the expected 3 blocks but returns `Ok(4)` and seeking to the very end of the file and reading returns `Err(SectorAlignment)`. "Quirks" like these are why I intended for this to be a bare-bones wrapper instead of supporting every feature the BIOS attempts to implement. I'll do some more tests and take a look at the BIOS function's disassembly, but it'll likely return `Result<()>` and `Ok(0)` will probably be relabeled as `Err`.

Oh wow. Just... wow. This BIOS is just so... I don't have words.

Maybe wrapping anything other than "IO a single block" is dangerous, then. In which case, the solution is simple: take an array reference. `arrayref` makes this ergonomic, sort of.

There could be slight performance advantages to IOing more than one block at a time from CD-ROM, but I'd guess if it matters that much for your application you'd want to program the controller directly rather than using the BIOS routines. You have to do that anyway for e.g. CDXA filtering or Red Book audio. And memory card IO via the BIOS is frightfully slow no matter how many blocks at a time you're doing...

> > An advantage of implementing Read/Write traits is that they allow adaptation
> 
> I agree, but keep in mind that with traits per filetype `MemCard` would effectively be `Write`. I can't imagine a situation here where `T: Write + !Read` or `T: Write + !Seek` is needed. I think `Read` might have some value, but the methods' behavior would depend on the filetype's block-size so it might not be so straightforward to implement `InflatingReader`. Since this module is intended to just wrap the BIOS calls in idiomatic rust instead of recreating every aspect of `std::fs`, I'd say let's avoid traits for now. Nothing really prevents external crates from defining and implementing these traits though.

Yeah. I suppose the same argument applies to a buffering adapter. Writing one is pretty trivial, and then the library user gets to tweak it to their own requirements.

> > Trying to use a generic adapter around a Read/Write implementation that chokes if the slice isn't exactly the right size would be a recipe for disaster
> 
> I think my explanation could've been better. This really isn't a problem since `read` with incorrectly-sized slices should _return_ `Err` instead of panicking. It's just a BIOS function call with a relabeled return code wrapped in `Result`. There shouldn't be wildly different behavior from what calling the BIOS from C would do (besides the call to `as_cstr` to allow input `&str` w/o null-termination).
> 
> > types CdromFile and MemCardFile that both implement BIOSFile
> 
> Like I mentioned earlier, since that trait would cover a subset of one type's methods this produces docs that are hard to follow. To be more concrete, the docs would show `write` under methods for `File<MemCard>` but `read` would be under `File<MemCard>`'s trait implementations and those methods aren't visible by default. Aside from the docs, conceptually it seems a bit strange to have some of the functionality (`write`) for one type of BIOS file outside the `BIOSFile` trait.

Yeah. Now that it's not 5AM and I'm less sick, I'm having an easier time following what you're saying here.

> > a lot of the time we won't be incurring any runtime overhead for it.
> 
> I understand the argument you're making, but I'd rather not add runtime checks based on the hope that the compiler will inline certain things.
> 
> > Oh, but this will be a cross-crate call, so I think that means it's less likely to be inlined without explicitly enabling lto?
> 
> To answer your question, cross-crate calls are not inlined without LTO. When marked with `#[inline]` (or even `#[inline(always)]`) they [_may_ be inlined](https://doc.rust-lang.org/reference/attributes/codegen.html#the-inline-attribute) without LTO, but are not guaranteed to be.

This is good to know.

**Comment by ayrtonm** [comment]
> Maybe wrapping anything other than "IO a single block" is dangerous, then. In which case, the solution is simple: take an array reference. arrayref makes this ergonomic, sort of.

An array reference is what `const_evaluatable` was doing. It's currently usable in only the simplest cases. For example I ran into issues trying to use this version of `write` to implement `putc` with a `&[u8; 128]`. arrayref could be an option for when `const_evaluatable` is stable, but it doesn't add anything without restricting the valid array reference sizes.

Also I just finished parameterizing  `File` by filetype and restricting methods that can be called on CD files. The comments still need to be tweaked,  but you can see how it's organized by generating the docs from the [fs branch](/ayrtonm/psx-sdk-rs/tree/fs) if you want to take a look.

**Comment by SolraBizna** [comment]
To clarify, what I meant was having only one `read`/`write` function that always IOs exactly one block. If the BIOS won't reliably report the length of partial IOs, then that's the only safe way to use these functions on block-based devices. This wouldn't require `const_evaluatable`, only the much simpler `min_const_generics`.

I haven't had time to look at or try the new `File` stuff yet, but I will within the next few days.

**Comment by SolraBizna** [comment]
I've had a look, and the documentation does seem pretty discoverable this way. An extra click is required to see the methods, but that's just down to how rustdoc works. The important things are that I had no trouble at all finding what methods were there, and in particular that it was very clear which methods were specific to `MemCard` files and which would work on any file.

It doesn't seem like there's much point to having `putc` or `getc` support for "blocky" files, though. There's no way to implement them correctly without buffering. The current implementation uses a whole block for each character, which I don't think is something anybody actually wants. (The BIOS implementations of `putc` and `getc` methods appear not to work with "blocky" files either.)

**Comment by ayrtonm** [comment]
> It doesn't seem like there's much point to having putc or getc support for "blocky" files

Yeah, I implemented them based on what the nocash specs say that the BIOS implementation does internally (i.e. redirects to `FileRead(fd,tempbuf,1)`) then realized that probably won't behave as expected. I might look at this when I get around to fixing the tests on the `fs` branch.

For now I'm focusing on adding some interrupt stuff I had previously removed to get CD files working. IIRC setting up IRQ handlers is required before calling `cd_init`.

**Comment by ayrtonm** [comment]
I fixed some issues in `File::read/write` and added a [demo for reading CD files](https://github.com/ayrtonm/psx-sdk-rs/blob/master/examples/fs/src/main.rs). `cargo psx run` won't work (as expected), but I had no problems running this with the ISOs produced by mkpsxiso.

**Comment by SolraBizna** [comment]
Finally got around to testing it. `getc` and `putc` working in 128-byte blocks and discarding 127 bytes seems weird to me (and `getc` won't work with CD-ROM files at all) but everything else looks and works fine.

Thanks!
